README.rst
===============================
Lichens for Analyzing XENON1T
===============================

.. image:: https://api.codacy.com/project/badge/Grade/724ba633bd6b4079b977e0aa623b327d
     :target: https://www.codacy.com/app/tunnell/lax?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=XENON1T/lax&amp;utm_campaign=Badge_Grade
     :alt: Style


This package contains the stanardized event selections for XENON1T analysis.


Documentation
--------------

Code organization
==================

Lax stores event selections in classes called "Lichens". These are organized in four different files:

* `sciencerun0`: selections for XENON1T's first (short) science run (SR0), using pax v.6.8.0 (NOT v6.4.2!);
* `sciencerun1`: selections for XENON1T's main science run (SR1), using pax v6.8.0;
* `postsr1`: improved versions of cuts applicable to the main science run (SR1). Applies to pax v6.8.0 OR pax v6.10.1; see the documentation of each cut for details.
* `sciencerun2`: selections for XENON1T's final science run (SR2), using pax v6.10.1.

The selections in the `sciencerun0` and `sciencerun1` files are exactly as they were used in analyses supporting XENON1T's main science paper (`Phys. Rev. Lett. 121, 111302 <https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.121.111302>`_ / `arXiv 1805.12562  <https://arxiv.org/abs/1805.12562>`_)



Applying selections
======================
The easiest way to apply selections is to use `hax.cuts.apply_lichen`, see the `second hax tutorial <https://github.com/XENON1T/hax/blob/master/examples/02_getting_serious.ipynb>`_ for more information. Alternatively, you can annotate the events (see below), then cut on the newly created columns.

Some lichens rely on the presence of non-standard variables in your event dataframe. This is usually indicated in the lichen docstrings. The default cut sets should work from the default set of minitrees


Annotating events
=======================
For some studies, you may wish to know which events would pass which selections. For this, use the `process` method of the individual lichen classes. For an example, see `here <https://github.com/XENON1T/lax/blob/master/examples/test_postsr1.ipynb>`_.



Credits
---------

Many XENON analysts contributed one or more event selections to lax: please see the lichen source files for details.

Development lead: Christopher Tunnell

Maintainers:

* Christopher Tunnell (SR0)
* Shayne Reichard (SR1, post-SR1)
* Patrick Deperio (SR1)
* Jelle Aalbers (post-SR1)


HISTORY.rst
=======
History
=======


1.7.2 (2019-05-10)
------------------
* S1PMT3fold cut for SR2 (#165)
* Preliminary S1AreaFractionTop cut (#164)


1.7.1 (2019-05-06)
------------------
* S2SingleScatter for SR2 (#160, #163)
* S1SingleScatter for SR2 (#161)
* Fix __init__.py for SR2 (#162)


1.7.0 (2019-03-28)
------------------
* Start SR2 lichen file (#157)

* SR1 (post-main paper):

  - CS2AreaFractiontopExtended updates (#156, #159)
  - MisIdSingleScatter (#154)

1.6.0 (2018-10-07)
------------------
* New postsr1 lichen file (#149)
* Issue and pull request templates (#151)
* CS2AreaFractionTopExtended (#152)
* S2PatternLikelihood extension to high-energies (#152)

1.5.2 (2018-05-15)
------------------
* Inner volume egg segmentation (#147)

1.5.1 (2018-02-27)
------------------
* Transpose, z-bin, and scaling bug fix, and smoothing of ER template in FV (#145, #146)

1.5.0 (2018-02-23)
------------------

* Add Z-optimized FV and swap in cut list (#144)

1.4.0 (2018-02-07)
------------------

* Fix MV cut; was not applied before (#129, #130)
* Small fix S1AreaUpperInjectionFraction line (#132)
* Tighten S1 width cut (#131)
* Use drift time instead of Z for S2 width cut (#134)
* Ignore dead PMTs and split top/bottom for S1PatternLikelihood (#137)
* New SingleElectronS2 cut (#139, #143)
* Use signal quantile of distance distribution only for PosDiff (#138)
* Switch to tighter corrected S2 AFT variable and use fitted quantile (#142)

1.3.0 (2018-01-05)
------------------

* MV cut to account for the coverage (#119)
* Remove MV cut from calibration data (#121)
* Use correct position for S1PatternLikelihood cut (#123)
* New S1 width cut to remove AC candidates (#124)
* New cut on area fraction near Rn220 injection points (#125)
* Tune S1 AFT cut based on bug fix and correct position (#126)
* Revert to SR0 version of S2 tail cut (#128)
* Optimise the PosDiff Cut (#127), interim update (#120)

1.2.2 (2017-12-04)
------------------

* Update S1PatternLikelihood Definition (#101)
* Add Flash cut to SR0 (#104)
* Add flash cut variable to SR1 (#102)
* Add cut for Misidentified krypton S1 (#103)
* Add check that nearest flash isn't nan (#106)
* Add parentheses (#107)
* Try again with flasher cut (#108)
* Remove S2 tail cut from calibration modes (#88)
* Try again Flash cut (#109)
* Update S2Width Cut (#96)
* Change bounds on S1SingleScatter Cut (#113)
* PosDiff Cut Definition (#111)
* S2Width Warnings Fix (#114)
* Add PosDiff; Remove S2Tails (duplicate); SR1 (#115)
* Remove S2Tails (duplicate); SR0 (#116)
* Fix S2Tails bug in cut lists (#117)


1.2.1 (2017-11-27)
------------------

* Muon Veto cut added (#55)
* Update on S1 Pattern Likelihood Cut (#99)
* Add super elliptical fiducial volumes for testing in SR1 (#100)

1.2.0 (2017-11-20)
------------------
Interim	release	for tracking data used in preliminary fits
(https://github.com/XENON1T/bbf/pull/35)
(https://github.com/XENON1T/SR1Results/pull/2)

* S1 area fraction top update and relegate to LowE (#56, #83)
* Update S2WidthCut and parameters for SR1 (#62, #75, #77)
* Fix LowEnergyRn220 accidentally removing DAQVeto cut (#66)
* Fix bug and treatment of 'nan' in S2SingleScatter (#69, #74)
* Reinstate InteractionPeaksBiggest (#71)
* Update S2Tail cut for SR1 (#76)
* S2PatternLikelihood Cut Tuning (#80, #87)
* Fix AmBe source position (#58, #97)
* Remove AmBe/NGFiducial cuts (#86, #98)
* Temporary 1.3T FV (#89, #94)
* Add Corrections minitree and SR1 switch to laxer (#64)
* Documentation (#73, #92)

1.0.0 (2017-06-22)
------------------

* Start SR1 lichens (#58)
* Neutron gun and updated AmBe FV (Also in #58)
* MC compatability (#57)
* Remove InteractionPeakBiggest for SR1 since done by pax

0.11.1 (2017-03-29)
------------------

* Refix SingleElectronS2 (#51)

0.11.0 (2017-03-29)
------------------

* Fix SingleElectronS2 (#50)

0.10.2 (2017-03-27)
------------------

* Bug in lichen definition (append not used as function)

0.10.0 (2017-03-24)
------------------

* Fix bug in SingleElectron lichen definition
* Create LowEnergyBackground, LowEnergyRn220, LowEnergyAmBe to simplfy cut application

0.9.5 (2017-03-23)
------------------

* sqrt(x*x + y*y) instead of r

0.9.4 (2017-03-10)
------------------

* Up version on S2 threshold

0.9.3 (2017-03-10)
------------------

* Raise S2 threshold to 200 PE

0.9.2 (2017-03-09)
------------------

* AmBe fiducial change to include z/r cut.
* Get runtime from hax for DAQ cut (#43)
* Fix warning in SingleElectronS2s (#44)

0.9.1 (2017-03-08)
------------------

* Fix to DAQ cut (#42)

0.9.0 (2017-03-08)
------------------

* DAQ cut update (#39)

0.8.6 (2017-03-08)
------------------

* Add larger FV option (#41)

0.8.5 (2017-03-07)
------------------

* Remove junk cut from main list since used for calibrations.

0.8.4 (2017-03-07)
------------------

* Fix bug in S1 width cut

0.8.3 (2017-03-07)
------------------

...

0.8.2 (2017-03-07)
------------------

* Fix bug in junk cut definition

0.8.1 (2017-03-07)
------------------

* Update cut list

lax 0.8.1

CutAllEnergy
  * CutFiducialCylinder1T v3
  * CutInteractionExists v0
  * CutS2Threshold v0
  * CutInteractionPeaksBiggest v0
  * CutS2AreaFractionTop v2
  * CutS2SingleScatter v2
  * CutDAQVeto v0
  * CutS1SingleScatter v1
  * CutS1AreaFractionTop v1
  * CutS2PatternLikelihood v0
CutLowEnergy
  * CutFiducialCylinder1T v3
  * CutS1LowEnergyRange v0
  * CutS2Threshold v0
  * CutInteractionPeaksBiggest v0
  * CutS2AreaFractionTop v2
  * CutS2SingleScatterSimple v0
  * CutDAQVeto v0
  * CutS1SingleScatter v1
  * CutS1AreaFractionTop v1
  * CutS2PatternLikelihood v0
  * CutS1PatternLikelihood v0
  * CutS2Width v2
  * CutS1MaxPMT v0
  * CutSignalOverPreS2Junk v1
  * CutSingleElectronS2s v0

0.8.0 (2017-03-07)
------------------

* Update area before main S2 < 300

0.7.0 (2017-03-07)
------------------

* Update 1T FV (#40)
* Tune S2 width cut (#38)
* S1 width cut (#33)
* S2 pattern likelihood (#34)
* S1 AFT speed fix (#32)

0.6.2 (2017-03-03)

lax/lichens/postsr1.py
"""Cuts for post-SR1 analyses.

The SR1 cuts are copied, unless they are overriden below.
"""
import inspect

import numpy as np

import lax
from lax.lichen import Lichen, ManyLichen, StringLichen  # pylint: disable=unused-import
from lax import __version__ as lax_version

from lax.lichens import sciencerun1 as sr1
DATA_DIR = sr1.DATA_DIR

from scipy import stats

# Import all lichens from sciencerun1
for x in dir(sr1):
    y = getattr(sr1, x)
    if inspect.isclass(y) and issubclass(y, lax.lichen.Lichen):
        locals()[x] = y

##
# Put new lichens here
##

class ERBandDEC(StringLichen):
    """A cut used in the double electron capture analysis to preselect the ER band
    Copied from xenon:xenon1t:analysis:backgrounds:ambe:fieguth:dec1t_final_analysis:cuts_overview

    Do NOT use at low energies (cS1 <200 PE)
    """
    version = 1
    string = "1 < log_cs_ratio < 2"

    def pre(self, df):
        df.loc[:, 'log_cs_ratio'] = np.log10(df['cs2']/df['cs1'])
        return df


class S2PatternLikelihood(StringLichen):
    """
    Extend S2 PatternLikelihood(S2 PLH) Cut up to 1.5e5 PE S2, which is good for up to around 220 keVee. This cut is a
    combination of SR1 S2PLH(s2 < 10000 PE) and extension of S2 PLH(1e4 < S2 < 1.5e5 PE), thus have same performance as 
    sr1 S2 PLH for low energy and have improved performance for high energy. S2 PLH cut aims to remove poorly 
    reconstructed events.

    The details can be found below:
    https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenon1t:jingqiang:darkphoton:s2patternlikelihood_200kev

    Requires Extended minitrees.
    Contact: Jingqiang Ye <jiy171@ucsd.edu>
    """
    version = 0
    p0 = (0.072, 594)
    p1_sr1 = (0.0404, 594, 0.0737, -686)


    string = ("((s2_pattern_fit < %.3f*s2 + %.3f) & (s2 > 10000))\
             | ((s2_pattern_fit < %.3f*s2 + %.3f*s2**%.3f + %.3f) & (s2 < 10000))" %(p0 + p1_sr1))


class CS2AreaFractionTopExtended(StringLichen):
    """"An extension of CS2AreaFractionTop to the entire S2 range
    with a designed acceptance of 99%.
    It is defined in the (cxys2, cs2_aft) space, with:

    cxys2 = (cs2_top + cs2_bottom) / s2_lifetime_correction
    cs2_aft = cs2_top / (cs2_top + cs2_bottom)

    Events where cxys2 > 2012700 PE or cxys2 < 90 PE are not cut.

    This cut should be used with a pax version of at least 6.10.0
    due to a major update of the S2 desaturation correction.

    Required minitrees: Corrections
    Defined with pax version: 6.10.1
    Wiki notes: xenon:xenon1t:double_beta:cs2_aft_extension
                xenon:xenon1t:double_beta:cs2_aft_extension_rejection_estimate
    Contact: Dominick Cichon <dominick.cichon@mpi-hd.mpg.de>"""

    version = 2

    top_bound_string = ('(6.306642E-01 + 2.032259E-07 * cxys2 +'
                        ' -5.450833E-13 * cxys2**2 + 6.182750E-19 * cxys2**3 +'
                        ' -3.087908E-25 * cxys2**4 + 5.637659E-32 * cxys2**5 +'
                        ' 2.123849E+00 / sqrt(cxys2) + -4.258579E+00 / cxys2)')
    bot_bound_string = ('(6.190286E-01 + 3.546500E-08 * cxys2 +'
                        ' -1.696261E-13 * cxys2**2 + 2.110380E-19 * cxys2**3 +'
                        ' -1.181304E-25 * cxys2**4 + 2.463864E-32 * cxys2**5 +'
                        ' -1.419829E+00 / sqrt(cxys2) + -5.642919E+00 / cxys2)')

    string = ('((' + top_bound_string + ' > cs2_aft) & (' + bot_bound_string +
              ' < cs2_aft)) | (cxys2 > 2012700) | (cxys2 < 90)')

    def pre(self, df):
        df.loc[:, 'cxys2'] = ((df['cs2_top'] + df['cs2_bottom']) /
                                df['s2_lifetime_correction'])
        df.loc[:, 'cs2_aft'] = df['cs2_top'] / (df['cs2_top'] +
                                                df['cs2_bottom'])
        return df


class CS2AreaFractionTopExtendedOldDesat(StringLichen):
    """"An extension of CS2AreaFractionTop to the entire S2 range
    with a designed acceptance of 99%.
    It is defined in the (cxys2, cs2_aft) space, with:

    cxys2 = (cs2_top + cs2_bottom) / s2_lifetime_correction
    cs2_aft = cs2_top / (cs2_top + cs2_bottom)

    Events where cxys2 > 2113500 PE or cxys2 < 90 PE are not cut.

    This is an alternate version of CS2AreaFractionTopExtended
    for the old S2 desaturation algorithm (pax versions older than 6.10.0).

    Required minitrees: Corrections
    Defined with pax version: 6.8.0
    Wiki notes: xenon:xenon1t:double_beta:cs2_aft_extension
                xenon:xenon1t:double_beta:cs2_aft_extension_rejection_estimate
    Contact: Dominick Cichon <dominick.cichon@mpi-hd.mpg.de>"""

    version = 1

    top_bound_string = ('(6.499452E-01 + 1.473286E-07 * cxys2 +'
                        ' -4.273597E-13 * cxys2**2 + 4.922129E-19 * cxys2**3 +'
                        ' -2.493411E-25 * cxys2**4 + 4.603164E-32 * cxys2**5 +'
                        ' 1.065584E+00 / sqrt(cxys2) + 5.840339E+00 / cxys2)')
    bot_bound_string = ('(6.184080E-01 + 3.561342E-09 * cxys2 +'
                        ' 2.792023E-14 * cxys2**2 + -1.133069E-19 * cxys2**3 +'
                        ' 7.043299E-26 * cxys2**4 + -1.242414E-32 * cxys2**5 +'
                        ' -1.388373E+00 / sqrt(cxys2) + -8.562613E+00 / cxys2)')

    string = ('((' + top_bound_string + ' > cs2_aft) & (' + bot_bound_string +
              ' < cs2_aft)) | (cxys2 > 2113500) | (cxys2 < 90)')

    def pre(self, df):
        df.loc[:, 'cxys2'] = ((df['cs2_top'] + df['cs2_bottom']) /
                                df['s2_lifetime_correction'])
        df.loc[:, 'cs2_aft'] = df['cs2_top'] / (df['cs2_top'] +
                                                df['cs2_bottom'])
        return df


class MisIdS1SingleScatter(Lichen):
    """Cut to target the shoulder on Kr83m data due to mis-identified krypton events.

    This cut is defined in the space of cs1 and largest_s2_before_main_s2_area.
    It's possible for one of the conversion electrons of the Kr83m decay to be mis-classified as an S2, 
    leading to an S2 that occurs in time before the main S2 that is larger than typical single-electrons. 
    Requires Extended minitrees.

    Note: xenon:shockley:misids1singlescatter
    Contact: Evan Shockley <ershockley@uchicago.edu>
    """

    version = 1.1
    
    pars = [60, 1.04, 4]  # from a fit to target only mis-Id Kr83m events
    s1_thresh = 155  # cs1 PE. Up to this value the cut will be a straight line
    cutval = 125 # straight line part of the cut
    min_s1 = 25 # smallest S1 to consider cutting
    
    
    def _cutline(self, x):
        return self.pars[0] + (self.pars[1] / 1e6) * (x - 220) ** self.pars[2]
    
    def cutline(self, x):
        return np.nan_to_num(self.cutval * (x < self.s1_thresh)) + np.nan_to_num((x >= self.s1_thresh) * self._cutline(x))
    
    def _process(self, df):
        df.loc[:, self.name()] = (np.nan_to_num(df.largest_s2_before_main_s2_area) < self.cutline(df.cs1)) | \
                                 (df.cs1 < self.min_s1)
        return df
